src/common/transformations/src/transformations/common_optimizations/skip_gather_before_transpose_and_reshape.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/skip_gather_before_transpose_and_reshape.hpp"
+#include "itt.hpp"
+
+#include <memory>
+#include <vector>
+
+#include <openvino/opsets/opset8.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+
+#include "transformations/utils/utils.hpp"
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::SkipGatherBeforeTransposeAndReshape, "SkipGatherBeforeTransposeAndReshape", 0);
+
+ngraph::pass::SkipGatherBeforeTransposeAndReshape::SkipGatherBeforeTransposeAndReshape() {
+    MATCHER_SCOPE(SkipGatherBeforeTransposeAndReshape);
+
+    auto input_m = ngraph::pattern::any_input(ngraph::pattern::has_static_dim(0));
+
+    auto indices_m = ngraph::pattern::wrap_type<ngraph::opset8::Constant>();
+    auto axis_m = ngraph::pattern::wrap_type<ngraph::opset8::Constant>();
+    auto gather_m = ngraph::pattern::wrap_type<ngraph::op::util::GatherBase>({input_m, indices_m, axis_m});
+
+    auto transpose_const_m = ngraph::pattern::wrap_type<ngraph::opset8::Constant>();
+    auto transpose_m = ngraph::pattern::wrap_type<ngraph::opset8::Transpose>({gather_m, transpose_const_m});
+
+    auto reshape_const_m = ngraph::pattern::wrap_type<ngraph::opset8::Constant>();
+    auto reshape_m = ngraph::pattern::wrap_type<ngraph::opset8::Reshape>({transpose_m, reshape_const_m});
+
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+        const auto& input = pattern_map.at(input_m);
+        if (input.get_partial_shape()[0] != 1) {
+            return false;
+        }
+
+        const auto gather = pattern_map.at(gather_m).get_node_shared_ptr();
+        const auto indices = as_type_ptr<ngraph::opset8::Constant>(pattern_map.at(indices_m).get_node_shared_ptr());
+        const auto axis = as_type_ptr<ngraph::opset8::Constant>(pattern_map.at(axis_m).get_node_shared_ptr());
+        if (!indices || !axis) {
+            return false;
+        }
+
+        const std::vector<std::int64_t> expected_gather_value{0};
+        if (indices->cast_vector<std::int64_t>() != expected_gather_value ||
+            axis->cast_vector<std::int64_t>() != expected_gather_value) {
+            return false;
+        }
+
+        const auto transpose = pattern_map.at(transpose_m).get_node_shared_ptr();
+        const auto transpose_const = as_type_ptr<ngraph::opset8::Constant>(pattern_map.at(transpose_const_m).get_node_shared_ptr());
+        if (!transpose_const) {
+            return false;
+        }
+
+        const auto reshape_const = as_type_ptr<ngraph::opset8::Constant>(pattern_map.at(reshape_const_m).get_node_shared_ptr());
+        if (!reshape_const) {
+            return false;
+        }
+
+        const auto reshape_vals = reshape_const->cast_vector<std::int64_t>();
```

Comment:
What about cases when reshape const remains on ShapeOf sub-graph? Looks like we will skip fusion, but we could try to get lower and upper bounds and check values for zero.
Example: ShapeOf(-1,-1,32,32)->Gather(axis=0, indices=[1,2,3])->Concat(axis=0 with Constant(value=[0]) - in this case you won't get a Constant for Reshape second input, but as we just want to check it for zero we can get bounds and theck them, so 0 (added by Concat) will be detected.

Reply:
@v-Golubev please confirm that this is no the case you have in your model, so we can merge this PR and implement this change later.